File: Python_TEA_model/economics_class.py
import numpy as np
import matplotlib.pyplot as plt


class economics_class:
    def __init__(self, inputs):
        get = inputs.get if isinstance(inputs, dict) else lambda key: getattr(inputs, key)

        self.lifetime = float(get("lifetime"))
        self.elec_price = float(get("elec_price"))
        self.inflation = float(get("inflation"))
        self.irr = float(get("irr"))
        self.debt_frac = float(get("debt_frac"))
        self.debt_IR = float(get("debt_IR"))
        self.tax_rate = float(get("tax_rate"))
        self.deprec = np.array(get("deprec"), dtype=float)
        self.annual_cost = np.array(get("annual_cost"), dtype=float)
        self.construc_IR = float(get("construc_IR"))
        self.OnM = float(get("OnM"))

        self.RROE = 0.0
        self.RINT = 0.0
        self.WACC = 0.0
        self.CRF = 0.0
        self.PVDEP = 0.0
        self.PFF = 0.0
        self.CFF = 0.0

        self.elec_price_hourly = None
        self.median_price = None
        self.charge_price = None
        self.discharge_price = None

        self.charge_cost_hourly = None
        self.discharge_revenue_hourly = None
        self.charge_cost = 0.0
        self.discharge_revenue = 0.0
        self.net_revenue = 0.0

        self.Ein = 0.0
        self.Eout = 0.0
        self.Qout = 0.0
        self.surface_capital_cost = 0.0
        self.subsurface_capital_cost = 0.0
        self.total_capital_cost = 0.0
        self.total_capital_cost_ITC = 0.0
        self.OnM_surface = 0.0
        self.OnM_subsurface = 0.0
        self.OnM_total = 0.0
        self.ITC = float(get("ITC", 0.0))
        self.elec_cost = 0.0
        self.FCR = 0.0
        self.LCOS = 0.0
        self.LCOE = 0.0
        self.LCOH = 0.0
        self.LCOS_ITC = 0.0
        self.LCOE_ITC = 0.0
        self.LCOH_ITC = 0.0

    def calc_fcr(self):
        self.RROE = (1 + self.irr) / (1 + self.inflation) - 1.0
        self.RINT = (1 + self.debt_IR) / (1 + self.inflation) - 1.0

        self.WACC = ((1 + self.inflation) * (1 + self.RROE) - 1) * (1 - self.debt_frac) + 1
        self.WACC = self.WACC + self.debt_frac * (
            ((1 + self.RINT) * (1 + self.inflation) - 1) * (1 - self.tax_rate)
        )
        self.WACC = self.WACC / (1 + self.inflation) - 1.0

        self.CRF = self.WACC / (1.0 - (1.0 / ((1 + self.WACC) ** self.lifetime)))

        self.PVDEP = 0.0
        for i, depreciation in enumerate(self.deprec, start=1):
            self.PVDEP += depreciation / (((1 + self.WACC) * (1 + self.inflation)) ** i)

        self.PFF = (1 - self.tax_rate * self.PVDEP) / (1.0 - self.tax_rate)

        self.CFF = 0.0
        for i, annual_cost in enumerate(self.annual_cost, start=1):
            self.CFF += (
                (((1 + self.construc_IR) ** (i - 0.5) - 1) * (1 - self.tax_rate) + 1)
                * annual_cost
            )

        self.FCR = self.CRF * self.PFF * self.CFF
        return self

    def calc_levelized_cost(self, mode: str):
        self.total_capital_cost = self.surface_capital_cost + self.subsurface_capital_cost
        self.total_capital_cost_ITC = (1.0 - self.ITC) * self.total_capital_cost
        self.OnM_total = self.OnM_surface + self.OnM_subsurface
        self.elec_cost = self.charge_cost if self.charge_cost > 0.0 else self.elec_price * self.Ein * 1e6

        if mode == "E":
            self.LCOE = (
                self.total_capital_cost * self.FCR
                + self.OnM_total
                + self.elec_cost
            ) / (self.Eout * 1e6)
            self.LCOE_ITC = (
                self.total_capital_cost_ITC * self.FCR
                + self.OnM_total
                + self.elec_cost
            ) / (self.Eout * 1e6)
        elif mode == "S":
            self.LCOS = (
                self.total_capital_cost * self.FCR
                + self.OnM_total
                + self.elec_cost
            ) / (self.Eout * 1e6)
            self.LCOS_ITC = (
                self.total_capital_cost_ITC * self.FCR
                + self.OnM_total
                + self.elec_cost
            ) / (self.Eout * 1e6)
        elif mode == "H":
            self.LCOH = (
                self.total_capital_cost * self.FCR
                + self.OnM_total
                + self.elec_cost
            ) / (self.Qout * 1e6)
            self.LCOH_ITC = (
                self.total_capital_cost_ITC * self.FCR
                + self.OnM_total
                + self.elec_cost
            ) / (self.Qout * 1e6)
        else:
            raise ValueError(f"Unsupported mode: {mode}")

        return self

    def calc_revenue(self, Win, Wout):
        self.charge_cost_hourly = np.multiply(self.elec_price_hourly, Win)
        self.discharge_revenue_hourly = np.multiply(self.elec_price_hourly, Wout)

        self.charge_cost = float(np.sum(self.charge_cost_hourly))
        self.discharge_revenue = float(np.sum(self.discharge_revenue_hourly))
        self.net_revenue = self.discharge_revenue - self.charge_cost
        return self
    
    '''
    def plot_revenue(self):
        if self.charge_cost_hourly is None or self.discharge_revenue_hourly is None:
            raise RuntimeError("Revenue hourly arrays are not available yet.")

        hours = np.arange(len(self.charge_cost_hourly))
        plt.figure(figsize=(12, 5))
        plt.plot(hours, self.charge_cost_hourly, label="Charge cost hourly")
        plt.plot(hours, self.discharge_revenue_hourly, label="Discharge revenue hourly")
        plt.xlabel("Hour")
        plt.ylabel("Cost / revenue")
        plt.legend()
        plt.tight_layout()
        plt.show()
    '''

File: MATLAB_models/economics_class.m
classdef economics_class
    properties
        lifetime
        elec_price
        inflation
        irr
        debt_frac
        debt_IR
        tax_rate
        deprec
        annual_cost
        construc_IR
        OnM
        ITC

        RROE
        RINT
        WACC
        CRF
        PVDEP
        PFF
        CFF

        elec_price_hourly
        median_price
        charge_price
        discharge_price

        charge_cost_hourly
        discharge_revenue_hourly
        charge_cost
        discharge_revenue
        net_revenue

        Ein % Annual Electricity input, GWh
        Eout % Annual Electricity output, GWh
        Qout % Annual heat output, GWh; for calculating LCOH
        surface_capital_cost
        subsurface_capital_cost
        total_capital_cost
        total_capital_cost_ITC
        OnM_surface
        OnM_subsurface
        OnM_total
        elec_cost
        FCR
        LCOS
        LCOE
        LCOH
        LCOS_ITC
        LCOE_ITC
        LCOH_ITC
    end

    methods
        function obj = economics_class(inputs)
            obj.lifetime    = inputs.lifetime ;      % Lifetime
            obj.elec_price  = inputs.elec_price ;      % Electricity price - dollars per kWhe. Lazard uses 0.033, ARPA-E uses 0.025. 0.06 is a value that I've used in the past.
            obj.inflation   = inputs.inflation ;      % Inflation
            obj.irr         = inputs.irr ;      % Internal Rate of Return
            obj.debt_frac   = inputs.debt_frac ;      % Project debt fraction - SAM is 0.60
            obj.debt_IR     = inputs.debt_IR ;      % Debt interest rate
            obj.tax_rate    = inputs.tax_rate ;      % Tax rate
            obj.deprec      = inputs.deprec ; % Depreciation[0.20 0.32 0.192 0.1152 0.1152 0.0576] ;
            obj.annual_cost = inputs.annual_cost; % Capital cost incurred in which years [0.80 0.10 0.10] ;
            obj.construc_IR = inputs.construc_IR ;         % Construction interest rate <- new assumption 25/1/17 to make CFF =1. SAM value -> % 0.08 ;
            obj.OnM         = inputs.OnM ;      % Operations and maintenance cost as a fraction of total capital cost - see Georgiou et al 2018
            obj.ITC         = inputs.ITC;       % Investment tax credit
        end

        % Calculate the fixed charge rate and other economic factors
        % Calculations based on the model in SAM
        function obj = calc_fcr(obj)
            obj.RROE = (1 + obj.irr) / (1 + obj.inflation) - 1.0 ;
            obj.RINT = (1 + obj.debt_IR) / (1 + obj.inflation) - 1.0 ;
            obj.WACC = ((1 + obj.inflation) * (1 + obj.RROE) - 1) * (1 - obj.debt_frac) + 1 ;
            obj.WACC = obj.WACC + obj.debt_frac * ((1 + obj.RINT) * (1 + obj.inflation) - 1) * (1 - obj.tax_rate) ;
            obj.WACC = obj.WACC / (1 + obj.inflation) - 1.0 ;

            obj.CRF  = obj.WACC / (1.0 - (1.0 / (1 + obj.WACC)^obj.lifetime) ) ;

            obj.PVDEP = 0.0 ;
            for i = 1:numel(obj.deprec)
                obj.PVDEP = obj.PVDEP + obj.deprec(i) / (((1 + obj.WACC) * (1 + obj.inflation))^i );
            end

            obj.PFF   = (1 - obj.tax_rate * obj.PVDEP) / (1.0 - obj.tax_rate) ;

            obj.CFF   = 0.0 ;
            for i = 1:numel(obj.annual_cost)
                obj.CFF = obj.CFF + (((1 + obj.construc_IR)^(i-0.5) - 1) * (1 - obj.tax_rate) + 1) * obj.annual_cost(i) ;
            end

            obj.FCR   = obj.CRF * obj.PFF * obj.CFF ;
        end


        % Calculate the levelised cost of energy
        function obj = calc_levelized_cost(obj,mode)

            obj.total_capital_cost = (obj.surface_capital_cost + obj.subsurface_capital_cost) ;
            obj.total_capital_cost_ITC = (1-obj.ITC) * obj.total_capital_cost;
            obj.OnM_total = obj.OnM_surface + obj.OnM_subsurface ;
            
            % If revenue previously calculated, use that for electricity cost
            if obj.charge_cost > 0
                obj.elec_cost = obj.charge_cost ;
            % Otherwise estimate from assumed electricity price
            else
                obj.elec_cost = obj.elec_price * obj.Ein * 1e6;
            end

            % Eout is total electricity generated per year, MWh-e
            switch mode
                case 'E'
                    obj.LCOE = (obj.total_capital_cost * obj.FCR + obj.OnM_total + obj.elec_cost) / (obj.Eout * 1e6) ;
                    obj.LCOE_ITC = (obj.total_capital_cost * obj.FCR * (1-obj.ITC) + obj.OnM_total + obj.elec_cost) / (obj.Eout * 1e6) ;
                case 'S'
                    obj.LCOS = (obj.total_capital_cost * obj.FCR + obj.OnM_total + obj.elec_cost) / (obj.Eout * 1e6) ;
                    obj.LCOS_ITC = (obj.total_capital_cost * obj.FCR * (1-obj.ITC) + obj.OnM_total + obj.elec_cost) / (obj.Eout * 1e6) ;
                case 'H'
                    obj.LCOH = (obj.total_capital_cost * obj.FCR + obj.OnM_total + obj.elec_cost) / (obj.Qout * 1e6) ;
                    obj.LCOH_ITC = (obj.total_capital_cost * obj.FCR * (1-obj.ITC) + obj.OnM_total + obj.elec_cost) / (obj.Qout * 1e6) ;
            end

            

        end

        function obj = calc_revenue(obj,Win,Wout)
            % Spending on electricity
            obj.charge_cost_hourly = obj.elec_price_hourly .* Win ;
            obj.discharge_revenue_hourly = obj.elec_price_hourly .* Wout ;

            obj.charge_cost = sum(obj.charge_cost_hourly);
            obj.discharge_revenue = sum(obj.discharge_revenue_hourly);
            obj.net_revenue = obj.discharge_revenue - obj.charge_cost ;
        end

    end

end